exporter/containerimage/export.go

```diff
@@ -267,7 +287,34 @@ func (e *imageExporterInstance) unpackImage(ctx context.Context, img images.Imag
 		return err
 	}
 
-	layers, err := getLayers(ctx, contentStore, manifest)
+	topLayerRef := src.Ref
+	if len(src.Refs) > 0 {
+		if r, ok := src.Refs[platforms.DefaultString()]; ok {
+			topLayerRef = r
+		} else {
+			return errors.Errorf("no reference for default platform %s", platforms.DefaultString())
+		}
+	}
+
+	remote, err := topLayerRef.GetRemote(ctx, true, e.layerCompression)
+	if err != nil {
+		return err
+	}
+
+	// ensure the content for each layer exists locally in case any are lazy
+	// TODO(sipsma) it would be cleaner to have an Applier that just used remote.Provider internally
+	eg, egctx := errgroup.WithContext(ctx)
+	for _, desc := range remote.Descriptors {
+		eg.Go(func() error {
+			return contentutil.Copy(egctx, contentutil.HashOnlyIngester(), remote.Provider, desc)
```

Comment:
Just to be clear, the point wasn't just to take the hash, it was to read the content in full from `remote.Provider` which had the (magic) side effect of de-lazying it into the local content store (due to the implementation in the cache package). However, now that you mention "double hashing" I do realize that when we do the magic de-lazy to the local content store, we are already doing a hash-verification, so copying to a `HashOnlyIngester` is actually unnecessary and I can at minimum change this to copy to an ingester that straight up discards data it reads without even taking the hash. I'll do that if nothing else.

But I also agree this all kind of sucks and the approach I suggested in the comment of just having the `Applier` use `remote.Provider` directly would be cleaner. However, I didn't see any way of doing so that didn't involve even more large-scale refactoring on top of the already large change here (happy to be wrong though). The problem is the actual `diff.Applier` interface only has this method
```
Apply(ctx context.Context, desc ocispec.Descriptor, mount []mount.Mount, opts ...ApplyOpt) (ocispec.Descriptor, error)
```
which doesn't include a `content.Provider`. The `Applier` implementation appears to internally use the `content.Store` it's hooked up to much earlier (i.e. [here](https://github.com/moby/buildkit/blob/488130002abbe5e372dda7192fd66d857ff93121/worker/containerd/containerd.go#L113) and [here](https://github.com/moby/buildkit/blob/488130002abbe5e372dda7192fd66d857ff93121/worker/runc/runc.go#L119)) as its `content.Provider`.

So the options I'm seeing besides what I already have are:
1. Wrap just the `Apply` method with something that takes the `desc` argument, de-lazies it (using a similar technique as now, just w/out the double-hashing you mentioned) and then calls the wrapped `Apply` method.
   * I have no opposition to this, I just didn't do it originally because it seemed like basically the same as what I'm currently doing but hidden under an `Apply` wrapper. Totally fine with updating to that if you think it's clearer though.
2. Change the original `Applier` to be constructed in the worker implementations with a special content store that can somehow be updated on the fly to use providers for given descriptors as set from the `remote.Provider` we obtain in the exporter here.
   * I'm not entirely sure how this would work, but the idea is appealing on some level. My main concern is that even if it's possible it would expand this PR way too much, but I could of course be overestimating and/or not seeing a simple path to implementing it.

Let me know if this makes any sense and what you think.

Reply:
> Wrap just the Apply method with something that takes the desc argument, de-lazies it (using a similar technique as now, just w/out the double-hashing you mentioned) and then calls the wrapped Apply method.

I think that is the same that I had in mind. 

`wrappedApplier(applier, remote/provider) Applier` implements `Apply` that when called invokes copy to be made to contentstore, and then just calls the regular applier that does the extraction from contentstore to snapshotter.

Reply:
If this turns out problematic for some reason(eg. the current parallelization issue) then lets add explicit method to ref/remote for this and avoid magic constructions with side-effects.

Reply:
Had some more thoughts; just to be super explicit about this since the code has kind of gone back and forth between different approaches here throughout the PR, here's what I'm thinking.

There's 3 places where a de-lazy may need to occur:
1. During `ImmutableRef.Extract()`
2. During a local export/unpack (such as the `Applier` case we were discussing)
3. Sometimes during a push; specifically when a lazy ref is going to be pushed to entire separate registry (unless there's some crazy way to do a cross-registry push, guessing not?)

There's also 2 general approaches for how the provider in the `Remote` returned by `ImmutableRef.GetRemote()` can be implemented:
1. The current approach w/ magic side-effects where the provider is implemented such that [opening a reader for it will result first in the entire descriptor blob to be pulled to the local content store and then a reader for the local content store to be returned](https://github.com/sipsma/buildkit/blob/94e8ab5a3b104d6ba7884375e2d8f391ceb3b1e5/cache/refs.go#L428-L482).
2. An alternative, less magical approach where it simply checks to see if the ref is lazy and, if so, set the provider to the one from the remote repo or, if not lazy, sets the local content store as the provider. This means that all callers of `GetRemote()` will have to check whether the descriptors in the returned remote are available locally before trying to use them.
   * Callers can be helped out with this via a `Unlazy` (or what I called before `EnsureContentExists`) method on the `Remote` or `ImmutableRef` object

In theory, approach 2 is preferable from the standpoint of being less magical and side-effecty, but there's a fairly significant downside when you consider the de-lazies that occur during a push (place 3 in the list above). Namely, all of the logic that determines during a push whether or not the content will need to exist locally or if pull can be skipped is fairly deep in external code (e.g. one place is [here where it's determined if you can do a cross-repo push](https://github.com/containerd/containerd/blob/0f2b15b7af41896190fcb6e7b597be789c11756b/remotes/docker/pusher.go#L129)). I'm guessing we can't modify that code to do the checks/calls required in approach 2, so the only other option would be to put all the logic that determines whether we can skip the local pull of content directly in the push util itself. That seems like it would most likely be a fairly large amount of copy/paste and hardcoding of behavior that should just be handled by the library we're calling out to.

So given that, I'm still inclined for approach 1 (though, like I said, I'll fix the whole double-hashing issue). However, I don't really see any reason why we couldn't do approach 1 but also add a util method to `Remote` called `Unlazy` that at least makes the code less confusing to read. I'll use `remote.Unlazy` in the wrapper around the `Applier` like we were talking about and also in the `extract()` code path. Internally, `Unlazy` will just do `contentutil.Copy` from the remote provider to an ingester that discards everything in order (without even checking hash) to trigger the magic side effect of de-lazying, but at least it will be wrapped in a more explicit method.

Let me know how that all sounds.

Reply:
> However, I don't really see any reason why we couldn't do approach 1 but also add a util method to Remote called Unlazy that at least makes the code less confusing to read. I'll use remote.Unlazy in the wrapper around the Applier like we were talking about and also in the extract() code path.

I agree. And if `Unlazy` is just a helper function that can be implemented currently (but it is hacky) we can just add it as a helper function `cache.Unlazy(cache.ImmutableRef)` so that it is more clear that it is just a nicer way to call existing properties of `ImmutableRef`.

>  Internally, Unlazy will just do contentutil.Copy from the remote provider to an ingester that discards everything

Internally Unlazy will call `ref.GetRemote().Provider.ReaderAt(ctx,ref.Desc)` that will copy from remote provider to the local contentstore?

Reply:
> Internally Unlazy will call ref.GetRemote().Provider.ReaderAt(ctx,ref.Desc) that will copy from remote provider to the local contentstore?

Almost but not exactly, due to the issues I mentioned before I think we have to leave `GetRemote` and the provider it returns [the same as it currently is](https://github.com/sipsma/buildkit/blob/94e8ab5a3b104d6ba7884375e2d8f391ceb3b1e5/cache/refs.go#L428-L482), where when the provider gets opened it first copies from the remote provider to the local content store before returning a reader for the local content store (de-lazies as a side-effect of opening the reader).

If it weren't for the case of de-lazies occurring during pushes that I mentioned in the previous comment, I think we could do what you suggest (`ref.GetRemote().Provider.ReaderAt(ctx,ref.Desc)`) where the `Remote` returned by `GetRemote` is just the actual remote provider w/out any de-lazying as a side-effect, but like I said before I just don't know how to make that work with the push handler code that exists outside this repo (and thus doesn't know how to call `Unlazy`).

So, given that, if `Unlazy` is going to be a helper function, it would have to do something like `contentutil.Copy(ctx, discardIngester, ref.GetRemote.Provider)`, where `discardIngester` is just the equivalent of `ioutil.Discard` but for ingesters.

However, if we make `Unlazy` be a method on `Remote` or `ImmutableRef`, we could probably make it not have to do the separate `contentutil.Copy`, it could be coded specially to just copy from the remote provider from `DescHandler` to the local content store. I think this is what I'd prefer since the `contentutil.Copy` is obviously sort of wasteful even if it's just writing to a discard ingester.

Reply:
> If it weren't for the case of de-lazies occurring during pushes that I mentioned in the previous comment, I think we could do what you suggest (ref.GetRemote().Provider.ReaderAt(ctx,ref.Desc)) where the Remote returned by GetRemote is just the actual remote provider w/out any de-lazying as a side-effect, 

I'm bit confused about what causes the conflict between push and de-lazy on `ref.GetRemote().Provider.ReaderAt` . The example you linked https://github.com/containerd/containerd/blob/0f2b15b7af41896190fcb6e7b597be789c11756b/remotes/docker/pusher.go#L129 directly works with descriptor and I don't see it requiring local data. So I don't think it would ever call `ReaderAt` if descriptor exists or can be cross-pushed. It only calls it if it needs local data and de-lazy is required then anyway.

Reply:
Right, what I'm saying is that there are times where de-lazy is required for push and there are times where it isn't required for push. However, the determination of whether it's needed happens deep in external code. So, for buildkit's code, if we don't have the "de-lazy as a side-effect" approach, then we would be forced to just always de-lazy before a push because we don't know ahead of time whether it needs to be present locally. But if we did that we'd be breaking some of the crucial features of lazy refs, namely that you don't have to de-lazy when just re-tagging or doing a cross-repo push.

Of course the other option would be to have buildkit's code do the determination of whether a de-lazy is needed. I was ruling that out for now because it seems like that functionality (such as that cross-repo logic I linked to) isn't public in containerd push handler code and is fairly far from trivial. Let me know though if there's some easier way of determining it, that would obviously change this consideration.

Reply:
We discussed this more on Slack, we were actually in agreement, I was just confused about one point. We’ll go with the side effect reader and Unlazy will use that internally, but it will just open the reader to trigger the side effect rather than copy to a discard ingester.

Reply:
Updated with this change. I did make one superficial tweak to it, namely, instead of having `Unlazy` be a util func, I made an `Unlazier` interface and then callers can check to see if a remote's provider implements that interface, calling it if so. The only small improvement over `Unlazy` as a util func is that you can avoid having to open and immediately close a reader to the local content store, otherwise it's basically the same in my mind. It's a fairly minor change so if you disagree w/ the approach just let me know and I have zero issue switching to `Unlazy` as a util func.